libraries/AP_OpticalFlow/AP_OpticalFlow_Linux.h

```diff
@@ -0,0 +1,84 @@
+/// -*- tab-width: 4; Mode: C++; c-basic-offset: 4; indent-tabs-mode: nil -*-
+
+/*
+ * Portions of this driver were borrowed from the PX4Firmware px4flow driver which can be found here:
+ *     https://github.com/PX4/Firmware/blob/master/src/drivers/px4flow/px4flow.cpp
+ */
+
+#ifndef AP_OpticalFlow_Linux_H
+#define AP_OpticalFlow_Linux_H
+
+#include "OpticalFlow.h"
+#include <AP_Math.h>
+
+class AP_OpticalFlow_Linux : public OpticalFlow_backend
+{
+public:
+    // constructor
+    AP_OpticalFlow_Linux(OpticalFlow &_frontend);
+
+    // initialise the sensor
+    void init();
+
+    // read latest values from sensor and fill in x,y and totals.
+    void update(void);
+
+private:
+
+    typedef struct {
+        uint16_t frame_count;
+        int16_t pixel_flow_x_sum;
+        int16_t pixel_flow_y_sum;
+        int16_t flow_comp_m_x;
+        int16_t flow_comp_m_y;
+        int16_t qual;
+        int16_t gyro_x_rate;
+        int16_t gyro_y_rate;
+        int16_t gyro_z_rate;
+        uint8_t gyro_range;
+        uint8_t sonar_timestamp;
+        int16_t ground_distance;
+    } i2c_frame;
+
+    typedef struct {
+        uint16_t frame_count_since_last_readout;
+        int16_t pixel_flow_x_integral;
+        int16_t pixel_flow_y_integral;
+        int16_t gyro_x_rate_integral;
+        int16_t gyro_y_rate_integral;
+        int16_t gyro_z_rate_integral;
+        uint32_t integration_timespan;
+        uint32_t sonar_timestamp;
+        uint16_t ground_distance;
+        int16_t gyro_temperature;
+        uint8_t qual;
+    } i2c_integral_frame;
```

Comment:
since you are using `sizeof(i2c_integral_frame)` to pass to `hal.i2c->readRegisters()` you are actually reading 3 bytes more than you should because this struct has a padding (`sizeof(i2c_integral_frame) == 28`, not `25`).  You need declare these structs as PACKED. Even if `i2c_frame` doesn't have this problem, I'd put a `PACKED` there, too.